File: cs/src/AutoDiff/ITermVisitor.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace AutoDiff
{
    /// <summary>
    /// Visitor for terms that has no result from its computations.
    /// </summary>
    public interface ITermVisitor
    {
        /// <summary>
        /// Performs an action for a constant term.
        /// </summary>
        /// <param name="constant">The input term.</param>
        void Visit(Constant constant);

        /// <summary>
        /// Performs an action for a zero term.
        /// </summary>
        /// <param name="zero">The input term.</param>
        void Visit(Zero zero);

        /// <summary>
        /// Performs an action for a constant power term.
        /// </summary>
        /// <param name="power">The input term.</param>
        void Visit(ConstPower power);

        /// <summary>
        /// Performs an action for a power term.
        /// </summary>
        /// <param name="power">The input term.</param>
        void Visit(TermPower power);

        /// <summary>
        /// Performs an action for a product term.
        /// </summary>
        /// <param name="product">The input term.</param>
        void Visit(Product product);

        /// <summary>
        /// Performs an action for a sum term.
        /// </summary>
        /// <param name="sum">The input term.</param>
        void Visit(Sum sum);
		
		/// <summary>
        /// Performs an action for a Gp term.
        /// </summary>
        /// <param name="sum">The input term.</param>
        void Visit(Gp gp);

        /// <summary>
        /// Performs an action for a variable term.
        /// </summary>
        /// <param name="variable">The input term.</param>
        void Visit(Variable variable);

        /// <summary>
        /// Performs an action for a logarithm term.
        /// </summary>
        /// <param name="log">The input term.</param>
        void Visit(Log log);

        /// <summary>
        /// Performs an action for an exponential function term.
        /// </summary>
        /// <param name="exp">The input term.</param>
        void Visit(Exp exp);
		
		///Additions by Carpe Noctem:
		
		/// <summary>
        /// Performs an action for a sigmoid term.
        /// </summary>
        /// <param name="sigmoid">The input term.</param>
        /// <returns>The result of the computation.</returns>
        void Visit(Sigmoid sigmoid);
		
		/// <summary>
        /// Performs an action for a sigmoid term.
        /// </summary>
        /// <param name="sigmoid">The input term.</param>
        /// <returns>The result of the computation.</returns>
        void Visit(LinSigmoid sigmoid);
		
		/// <summary>
        /// Performs an action for a less-than constraint.
        /// </summary>
        /// <param name="constraint">The input term.</param>
        /// <returns>The result of the computation.</returns>
        void Visit(LTConstraint constraint);	
		
		/// <summary>
        /// Performs an action for a less-than-or-equal constraint.
        /// </summary>
        /// <param name="constraint">The input term.</param>
        /// <returns>The result of the computation.</returns>
        void Visit(LTEConstraint constraint);	
		
		/// <summary>
        /// Performs an action for a min term.
        /// </summary>
        /// <param name="min">The input term.</param>
        /// <returns>The result of the computation.</returns>
        void Visit(Min min);
		
		
		/// <summary>
        /// Performs an action for a max term.
        /// </summary>
        /// <param name="max">The input term.</param>
        /// <returns>The result of the computation.</returns>
        void Visit(Max max);		
		
		/// <summary>
        /// Performs an action for an and term.
        /// </summary>
        /// <param name="min">The input term.</param>
        /// <returns>The result of the computation.</returns>
        void Visit(And and);
		
		/// <summary>
        /// Performs an action for an or term.
        /// </summary>
        /// <param name="min">The input term.</param>
        /// <returns>The result of the computation.</returns>
        void Visit(Or or);
		
		/// <summary>
        /// Performs an action for a ConstraintUtility term.
        /// </summary>
        /// <param name="cu">The input term.</param>
        /// <returns>The result of the computation.</returns>
        void Visit(ConstraintUtility cu);	
       
		 /// <summary>
        /// Performs an action for a sine term.
        /// </summary>
        /// <param name="sin">The input term.</param>
        /// <returns>The result of the computation.</returns>
        void Visit(Sin sin);		

        /// <summary>
        /// Performs an action for a cosine term.
        /// </summary>
        /// <param name="cos">The input term.</param>
        /// <returns>The result of the computation.</returns>
        void Visit(Cos cos);		

        /// <summary>
        /// Performs an action for an abs term.
        /// </summary>
        /// <param name="abs">The input term.</param>
        /// <returns>The result of the computation.</returns>
        void Visit(Abs abs);	
		
		/// <summary>
		/// Performs an action for an atan2 term.
		/// </summary>
		/// <param name='atan2'>
		/// Atan2.
		/// </param>
		void Visit(Atan2 atan2);
		
		/// <summary>
		/// Performs an action for a reified constraint
		/// </summary>
		/// <param name='r'>
		/// R.
		/// </param>
		void Visit(Reification r);
		
		
    }
}

File: java/src/main/java/AutoDiff/ITermVisitor.java
//
// Translated by CS2J (http://www.cs2j.com): 12.11.2016 09:40:33
//

package AutoDiff;

/**
* A visitor for the terms that has a result from its computations
* The type of the computation results
*/
public interface ITermVisitor <TResult>  
{
    /**
    * Computes a value for a constant term.
    * 
    *  @param constant The input term.
    *  @return The result of the computation.
    */
    TResult visit(Constant constant) throws Exception ;

    /**
    * Computes a value for a zero term.
    * 
    *  @param zero The input term.
    *  @return The result of the computation.
    */
    TResult visit(Zero zero) throws Exception ;

    /**
    * Computes a value for a power term.
    * 
    *  @param power The input term.
    *  @return The result of the computation.
    */
    TResult visit(ConstPower power) throws Exception ;

    /**
    * Computes a value for a power term.
    * 
    *  @param power The input term.
    *  @return The result of the computation.
    */
    TResult visit(TermPower power) throws Exception ;

    /**
    * Computes a value for a product term.
    * 
    *  @param product The input term.
    *  @return The result of the computation.
    */
    TResult visit(Product product) throws Exception ;

    /**
    * Computes a value for a sum term.
    * 
    *  @param sum The input term.
    *  @return The result of the computation.
    */
    TResult visit(Sum sum) throws Exception ;

    /**
    * Computes a value for a variable term.
    * 
    *  @param variable The input term.
    *  @return The result of the computation.
    */
    TResult visit(Variable variable) throws Exception ;

    /**
    * Computes a value for a logarithm term.
    * 
    *  @param log The input term.
    *  @return The result of the computation.
    */
    TResult visit(Log log) throws Exception ;

    /**
    * Computes a value for an exponential function term.
    * 
    *  @param exp The input term.
    *  @return The result of the computation.
    */
    TResult visit(Exp exp) throws Exception ;

    /**
    * Additions by Carpe Noctem:
    * 
    * Computes a value for a sigmoid term.
    * 
    *  @param sigmoid The input term.
    *  @return The result of the computation.
    */
    TResult visit(Sigmoid sigmoid) throws Exception ;

    /**
    * Computes a value for a sigmoid term.
    * 
    *  @param sigmoid The input term.
    *  @return The result of the computation.
    */
    TResult visit(LinSigmoid sigmoid) throws Exception ;

    /**
    * Computes a value for a less-than constraint term.
    * 
    *  @param constraint The input term.
    *  @return The result of the computation.
    */
    TResult visit(LTConstraint constraint) throws Exception ;

    /**
    * Computes a value for a less-than-or-equal constraint term.
    * 
    *  @param constraint The input term.
    *  @return The result of the computation.
    */
    TResult visit(LTEConstraint constraint) throws Exception ;

    /**
    * Computes a value for a min term.
    * 
    *  @param sigmoid The input term.
    *  @return The result of the computation.
    */
    TResult visit(Min min) throws Exception ;

    /**
    * Computes a value for a max term.
    * 
    *  @param max The input term.
    *  @return The result of the computation.
    */
    TResult visit(Max max) throws Exception ;

    /**
    * Computes a value for an and term.
    * 
    *  @param sigmoid The input term.
    *  @return The result of the computation.
    */
    TResult visit(And and) throws Exception ;

    /**
    * Computes a value for an or term.
    * 
    *  @param max The input term.
    *  @return The result of the computation.
    */
    TResult visit(Or or) throws Exception ;

    /**
    * Computes a value for a ConstraintUtility term.
    * 
    *  @param cu The input term.
    *  @return The result of the computation.
    */
    TResult visit(ConstraintUtility cu) throws Exception ;

    /**
    * Computes a value for a sine term.
    * 
    *  @param sin The input term.
    *  @return The result of the computation.
    */
    TResult visit(Sin sin) throws Exception ;

    /**
    * Computes a value for a cosine term.
    * 
    *  @param cos The input term.
    *  @return The result of the computation.
    */
    TResult visit(Cos cos) throws Exception ;

    /**
    * Computes a value for an exponential function term.
    * 
    *  @param abs The input term.
    *  @return The result of the computation.
    */
    TResult visit(Abs abs) throws Exception ;

    TResult visit(Atan2 atan2) throws Exception ;

    TResult visit(Reification r) throws Exception ;

    TResult visit(Negation r) throws Exception ;

}


